### PR TITLE
masking compatible with fullgraph compile

### PR DESCRIPTION
**Pull request by theAdamColton**
this adds some slightly confusing masking code, but improves speed by 3x by making the shape of intermediate tensors non-dynamic. The masked_mean code is equivalent, up to fp precision, with the old code that used tensor indexing

Before, using LFQ with masking was not compatible with torch.compile with fullgraph=True or with dynamic=False. It was compatible with plain torch.compile, but the masked tensor indexing caused graph breaks

I added an example that uses masked sequences, to make sure it works properly

I did a benchmark. I ran the example code that uses masking. This was on a 3090 GPU
- the previous masked LFQ implementation, using torch.compile(model, fullgraph=False, mode='max-autotune'), had an average model.forward time of 1.18 milliseconds
- with this commit, using torch.compile(model, fullgraph=True, mode='max-autotune'), the average time is 0.40 milliseconds

The speedup might be worth the extra confusingness in the code

### CODE REVIEW

**Comment by lucidrains** [comment]
ah yea, that does look a bit confusing, needs a tiny bit more work

do you think you can try fitting all the logic into one function, `masked_mean`, where if `mask` is `None`, it simply takes a regular `.mean()`?

**Comment by lucidrains** [comment]
we can reassess after your refactor

**Comment by lucidrains** [comment]
@theAdamColton have you tried the updated LFQ? curious how you got good results on the previous broken one

**Comment by theAdamColton** [comment]
With the previous LFQ i set entropy loss and commit loss to very low weights and it did actually work.

**Comment by theAdamColton** [comment]
I've also been experimenting with the entropy loss [from maskgit](https://github.com/google-research/maskgit/blob/1db23594e1bd328ee78eadcd148a19281cd0f5b8/maskgit/libml/losses.py#L190), it does it slightly different than the current lfq code here. The one there seems to work pretty well

**Comment by theAdamColton** [comment]
Also, this is a different issue, but I think [here](https://github.com/lucidrains/vector-quantize-pytorch/blob/4fc8e0a8f032d0eaeb074eb8a71edd19f8b624ec/vector_quantize_pytorch/lookup_free_quantization.py#L218C1-L218C1) where the entropy is computed, maybe it should use F.log_softmax to separately compute the log probs from the distances, instead of taking the log of the probs to get the log probs.

**Comment by lucidrains** [comment]
> Also, this is a different issue, but I think [here](https://github.com/lucidrains/vector-quantize-pytorch/blob/4fc8e0a8f032d0eaeb074eb8a71edd19f8b624ec/vector_quantize_pytorch/lookup_free_quantization.py#L218C1-L218C1) where the entropy is computed, maybe it should use F.log_softmax to separately compute the log probs from the distances, instead of taking the log of the probs to get the log probs.

@theAdamColton how is that different? can you show me in code?

**Comment by theAdamColton** [comment]
@lucidrains 
for example, instead of
```
prob = (-distance * inv_temperature).softmax(dim = -1)
per_sample_entropy = (-prob * log(prob)).sum(dim=-1).mean()
```
this is what I mean:
```
prob = (-distance * inv_temperature).softmax(dim = -1)
log_prob = F.log_softmax(-distance * inv_temperature, dim = -1)
per_sample_entropy = (-prob * log_prob).sum(dim=-1).mean()
```

I don't know if it would make a difference, but it's what the maskgit code does. Using log_softmax might fix precision issues

[from the pytorch log_softmax doc ](https://pytorch.org/docs/stable/generated/torch.nn.functional.log_softmax.html)
"While mathematically equivalent to log(softmax(x)), doing these two operations separately is slower and numerically unstable. This function uses an alternative formulation to compute the output and gradient correctly."

**Comment by lucidrains** [comment]
I think the numerical stability is accounted for by the epsilon in the log I have in the file, but do let me know otherwise

**Comment by lucidrains** [comment]
anyways, I've put in my hours today, happy Saturday!
